### PR TITLE
Remove Job from PS API

### PR DESCRIPTION
**Pull request by duglin**
One step closer to removing engine.Table too  :-)

In case people wonder, the code at: https://github.com/docker/docker/compare/master...duglin:RemoveJobPS?expand=1#diff-44a527ea72ea9572e1cf71c7f1e2a0f0L401  appears to be broken based on testing that @jfrazelle and I did.  It always return an empty list and in order for that if-statement to even be executed it requires a CLI to use v1.4 API - which is pretty darn old - way way pre-Docker 1.0.0.   I think its safe to remove it.

Signed-off-by: Doug Davis dug@us.ibm.com

### CODE REVIEW

**Comment by jfrazelle** [comment]
I can confirm that using a new daemon v1.18 and a 0.6.0 client that this does not work anyways, so +1 to removing

**Comment by jfrazelle** [comment]
But if anyone else comes here and wonders the client version 0.7.0 works perfectly, its kinda like the coolest thing ever

**Comment by duglin** [comment]
ping @tiborvass @LK4D4

**Comment by duglin** [comment]
ping @icecrime

File: api/server/server.go [hunk 1]
```diff
@@ -374,42 +374,31 @@ func getContainersTop(eng *engine.Engine, version version.Version, w http.Respon
...
-	job.Setenv("filters", r.Form.Get("filters"))
 
-	if version.GreaterThanOrEqualTo("1.5") {
-		streamJSON(job, w, false)
-	} else if outs, err = job.Stdout.AddTable(); err != nil {
-		return err
-	}
-	if err = job.Run(); err != nil {
-		return err
+	config := daemon.ContainersConfig{
```
Comment (LK4D4) [inline api/server/server.go]: I don't know why, but I prefer to create pointer variable directly here, because config as value isn't used anyway.

File: api/server/server.go [hunk 2]
```diff
@@ -374,42 +374,31 @@ func getContainersTop(eng *engine.Engine, version version.Version, w http.Respon
...
-		for _, out := range outs.Data {
-			ports := engine.NewTable("", 0)
-			ports.ReadListFrom([]byte(out.Get("Ports")))
-			out.Set("Ports", api.DisplayablePorts(ports))
-		}
-		w.Header().Set("Content-Type", "application/json")
-		if _, err = outs.WriteListTo(w); err != nil {
+
+	if tmp := r.Form.Get("limit"); tmp != "" {
+		if config.Limit, err = strconv.Atoi(tmp); err != nil {
```
Comment (LK4D4) [inline api/server/server.go]: I feel bad about such flow. I prefer to keep local variables here. Use:
```
limit, err := ...
if err != nil {
    return err
}
config.Limit = limit
```

File: api/server/server.go [hunk 3]
```diff
@@ -374,42 +374,31 @@ func getContainersTop(eng *engine.Engine, version version.Version, w http.Respon
...
-	if version.LessThan("1.5") { // Convert to legacy format
-		for _, out := range outs.Data {
-			ports := engine.NewTable("", 0)
-			ports.ReadListFrom([]byte(out.Get("Ports")))
-			out.Set("Ports", api.DisplayablePorts(ports))
-		}
-		w.Header().Set("Content-Type", "application/json")
-		if _, err = outs.WriteListTo(w); err != nil {
+
+	if tmp := r.Form.Get("limit"); tmp != "" {
```
Comment (LK4D4) [inline api/server/server.go]: `tmp` isn't nice variable name. maybe `fl` or `l` or you can use `r.Form.Get("limit") twice.

File: api/server/server.go (hunk 1 above)
Comment (duglin) [inline api/server/server.go]: ok done

File: daemon/list.go [hunk 4]
```diff
@@ -23,35 +21,41 @@ func (daemon *Daemon) List() []*Container {
...
 func (r ByCreated) Len() int           { return len(r) }
 func (r ByCreated) Swap(i, j int)      { r[i], r[j] = r[j], r[i] }
 func (r ByCreated) Less(i, j int) bool { return r[i].Created < r[j].Created }
 
-func (daemon *Daemon) Containers(job *engine.Job) error {
+func (daemon *Daemon) Containers(config *ContainersConfig) ([]types.Container, error) {
 	var (
 		foundBefore bool
 		displayed   int
-		all         = job.GetenvBool("all")
```
Comment (LK4D4) [inline daemon/list.go]: ah, so nice that we removed so much vars :) Always hated them.

File: api/server/server.go (hunk 2 above)
Comment (duglin) [inline api/server/server.go]: See if you're ok with the what I have now.  I don't like using errors/exceptions to detect valid flow/data.

File: api/server/server.go (hunk 3 above)
Comment (duglin) [inline api/server/server.go]: see latest code to see if you're ok with that.

**Comment by LK4D4** [comment]
I'm okay with this. But I should say that for events I moved all filtering to api. So my daemon method just returning events. Maybe we should do same for `daemon.Containers()`?

**Comment by duglin** [comment]
I think it depends on who should own filtering.  Let's say we have multiple transports, and not just http, would we want to duplicate that logic in each transport/API layer or allow the core daemon to do it in just one spot?  If the logic isn't specific to the transport then I tend to want to push it down so its common code and can be reused.

**Comment by LK4D4** [comment]
@duglin it can be done by function `filterConts([]*daemon.Containers, since....)` outside of daemon. I'm not sure that filtering belongs to daemon package.

**Comment by duglin** [comment]
I understand that it can be done - its just a question of whether we want to make the caller do it as a second step or make it part of the interface to retrieve the list.  If the list is very large I'm sure they would appreciate it being filtered before it got to them.

**Comment by LK4D4** [comment]
@duglin I'm not sure why they will appreciate filtering if it is pressure on garbage collector. List size is not an issue here at all.
I personally think that I want to make the caller do filtering, because why not? Decomposition and readability vs god objects/functions. I choose first.
Okay I think for this PR it is easier just to left it as it is.

File: daemon/list.go [hunk 5]
```diff
@@ -23,35 +21,41 @@ func (daemon *Daemon) List() []*Container {
...
+}
+
 type ByCreated []types.Container
 
 func (r ByCreated) Len() int           { return len(r) }
 func (r ByCreated) Swap(i, j int)      { r[i], r[j] = r[j], r[i] }
 func (r ByCreated) Less(i, j int) bool { return r[i].Created < r[j].Created }
 
-func (daemon *Daemon) Containers(job *engine.Job) error {
+func (daemon *Daemon) Containers(config *ContainersConfig) ([]types.Container, error) {
```
Comment (LK4D4) [inline daemon/list.go]: `types.Container` is superbig with slices and maps, makes sense to make it `[]*types.Container`.

File: daemon/list.go [hunk 6]
```diff
@@ -197,14 +201,11 @@ func (daemon *Daemon) Containers(job *engine.Job) error {
 	for _, container := range daemon.List() {
 		if err := writeCont(container); err != nil {
 			if err != errLast {
-				return err
+				return nil, err
 			}
 			break
 		}
 	}
 	sort.Sort(sort.Reverse(ByCreated(containers)))
```
Comment (LK4D4) [inline daemon/list.go]: I think this is not necessary because `daemon.List` should be already sorted.

**Comment by duglin** [comment]
yes - let's not hold up this PR for this bike-shedding conversation but just to explain my thinking....

on IRC you said: also, now without jobs we're creating superbig slice of filtered containers in addition to superbig slice of daemon.container.List()

I think its important to separate out our particular implementation choices today from the interfaces.  You're correct that right now we create superbig slices of all containers and then filter it down to another superbig slice.  But that's because of our choices. If we stored our container info into a DB and applied the filters as we retrieved the list of containers then there's only one superbig slice involved.

If we did what you're suggesting, then we'd always return the superbig slice back to the caller not matter what - even if the net result was just 5.  In my preferred solution, no slice would ever be larger than 5. And we can get there by introducing a DB and the caller would never know that we've improved - aside from memory consumption/time.  In what you're proposing, they would never benefit from the DB switch because they're always going to get superbig results even when the filter brings it down to 5.

Doing this on the list of containers isn't probably the best example because the list isn't as large as say events where we could be talking about Gigs of data.  The less we move that around the better, so the sooner we can apply the filters the better. And applying it to a DB query, rather than leaving it to the caller, would be more performant, IMO.

File: daemon/list.go (hunk 6 above)
Comment (duglin) [inline daemon/list.go]: hmmm can you double check... I agree daemon.List is sorted, but this wants it reversed.

Comment (duglin) [inline daemon/list.go]: for posterity... lk4d4 and I checked and this sort() is unnecessary.

File: daemon/list.go [hunk 7]
```diff
@@ -23,35 +20,41 @@ func (daemon *Daemon) List() []*Container {
...
+type ContainersConfig struct {
+	All     bool
+	Since   string
+	Before  string
+	Limit   int
+	Size    bool
+	Filters string
+}
+
+type ByCreated []*types.Container
```
Comment (LK4D4) [inline daemon/list.go]: Do we still need this type?

File: api/server/server.go (hunk 2 above)
Comment (LK4D4) [inline api/server/server.go]: I didn't understand what you don't like :) Could you rephrase? :blush:

File: daemon/list.go (hunk 7 above)
Comment (duglin) [inline daemon/list.go]: Ah good catch - removed!

File: api/server/server.go (hunk 2 above)
Comment (duglin) [inline api/server/server.go]: If I understood you correctly, you wanted the code to allow for "" to be passed into Itoa(), and let it generate an error, rather than have an explicit check for "" before the Itoa() call.  I was just saying that I I'm not fond of having errors generated in the case of valid input, when possible.  Since "" is a valid value for `limit` I would prefer to check for it prior to letting Itoa() generate an error.  Not a big deal, just a preference I have.

Comment (LK4D4) [inline api/server/server.go]: ah, no, I meant to not use `err` variable from outer scope and not assign value to `config.Limit` directly, because you know `:=/=` confusions is pretty common. And here you don't need outer scope and I think that local scope easier to read. So, my snippet meant to be inside `if != ""`.

Comment (duglin) [inline api/server/server.go]: Are you ok with what’s in there now?
https://github.com/docker/docker/pull/12163/files#diff-44a527ea72ea9572e1cf71c7f1e2a0f0R390
-Doug

Comment (LK4D4) [inline api/server/server.go]: No, I want:
```
	if tmpLimit := r.Form.Get("limit"); tmpLimit != "" {
		limit, err := strconv.Atoi(tmpLimit)
		if err != nil {
 			return err
 		}
                config.Limit = limit
        }
```
and removing `var err error`

Comment (duglin) [inline api/server/server.go]: I don’t see the point in the tmp variable when we can assign it directly to config.Limit and if there’s an error that entire config struct vanishes anyway.
-Doug

Comment (LK4D4) [inline api/server/server.go]: Point is not to use outer scope `err`. Local scope always better reads, and `var err error` is always scary, because you need to look through whole code to see for what it needed - sometimes it is necessary, but not here.
Also all over the code used `val, err := func()` pattern, and I prefer to use less `globalVar, globalErr = func()` pattern.
Another style concern that `config.Limit` is global and `err` local by logic and mixing them in same assignment sorta tricky.

Comment (duglin) [inline api/server/server.go]: Sorry I just don’t see the point in a new variable when I think the code is clear as it is.

**Comment by crosbymichael** [comment]
LGTM

File: api/server/server.go (hunk 2 above)
Comment (LK4D4) [inline api/server/server.go]: It is matter of style. If you don't see - okay.
Your code of course clear for you.

**Comment by duglin** [comment]
good to merge?

**Comment by icecrime** [comment]
I reviewed without reading through all the debates. I dislike the `var err error` too but this PR overall does much more good than harm: LGTM :+1:

File: api/server/server.go [hunk 8]
```diff
@@ -374,42 +374,32 @@ func getContainersTop(eng *engine.Engine, version version.Version, w http.Respon
...
+		All:     r.Form.Get("all") == "1",
+		Size:    r.Form.Get("size") == "1",
+		Since:   r.Form.Get("since"),
+		Before:  r.Form.Get("before"),
+		Filters: r.Form.Get("filters"),
 	}
-	if err = job.Run(); err != nil {
-		return err
-	}
-	if version.LessThan("1.5") { // Convert to legacy format
```
Comment (HuKeping) [inline api/server/server.go]: @jfrazelle your test break probably because missing the nil check of var `outs` here :), anyway it's no need to fix this  bug now since the change of this PR.
